### DISCUSSION TITLE
kickoff() Is it thread safe？

### DISCUSSION THREAD

**Discussion by tianshangwuyun** [crewAIInc/crewAI#1569] [Q&A]
I want to know if this method is thread safe, and if the state in the flow is stored in memory and thread safe

**Comment by lakshit2234** [comment]
`kickoff()` is **not thread-safe** by default in CrewAI.

Here's why:

**State storage:** The flow state is stored in-memory as a Python object 
on the `Flow` instance. If multiple threads share the same `Flow` instance 
and call `kickoff()` simultaneously, they will read/write the same state 
object — causing race conditions.

**Safe approaches:**

1. **One instance per thread** — create a separate `Flow` instance for 
each thread/request. This is the recommended pattern:

```python
import threading
from crewai.flow.flow import Flow

def run_flow():
    flow = MyFlow()  # new instance per thread
    flow.kickoff()

threads = [threading.Thread(target=run_flow) for _ in range(5)]
for t in threads:
    t.start()
```

2. **Use async instead** — `kickoff_async()` with `asyncio` is safer 
for concurrent execution since each coroutine has its own execution context.

3. **Avoid shared mutable state** — if you must share a `Flow` instance, 
protect state mutations with `threading.Lock()`.

**Bottom line:** treat each `Flow` instance as single-use and non-shared 
across threads. Instantiate fresh per request/thread for safe concurrent usage.
